### ISSUE TITLE
$0 in extract function assist

### ISSUE THREAD

**Issue by weigangd** [rust-analyzer/rust-analyzer#7793]
Hello, I am using nvim + LanguageClient-neovim and I have no plugin for snippet support. When I use the "extract function assist" code action, the generated function always has the name $0fun_name. The same happens with nvim-lsp. If I understand #2518 and #4117 correctly, snippet-completion should be disabled automatically if the client does not support snippets.

**Comment by Veykril** [comment]
Looks like we don't honor the snippet capabilities in that assist https://github.com/rust-analyzer/rust-analyzer/blob/d5b8a401f38f88ecb0d894c88dcbab3d338a0ef2/crates/ide_assists/src/handlers/extract_function.rs#L1071-L10892

**Comment by Veykril** [comment]
The fix is to check for the capability like done in `add_missing_impl_members` assist here https://github.com/rust-analyzer/rust-analyzer/blob/d5b8a401f38f88ecb0d894c88dcbab3d338a0ef2/crates/ide_assists/src/handlers/add_missing_impl_members.rs#L121-L140 and add the snippet `$0` conditionally.

**Comment by goodSyntax808** [comment]
@Veykril  you mean like this? That simple? Simply return `""` if there isn't that capability, otherwise, do what it already does?

```rust
fn format_function(
    ctx: &AssistContext,
    module: hir::Module,
    fun: &Function,
    old_indent: IndentLevel,
    new_indent: IndentLevel,
) -> String {
    let mut fn_def = String::new();

    match ctx.config.snippet_cap {
        None => (),
        Some(_) => {
            let params = make_param_list(ctx, module, fun);
            let ret_ty = make_ret_ty(ctx, module, fun);
            let body = make_body(ctx, old_indent, new_indent, fun);
            format_to!(fn_def, "\n\n{}fn $0{}{}", new_indent, fun.name, params);
            if let Some(ret_ty) = ret_ty {
                format_to!(fn_def, " {}", ret_ty);
            }
            format_to!(fn_def, " {}", body);
        }
    };

    fn_def
}
```

**Comment by Veykril** [comment]
No not quite, we still want to emit the code but dont want to emit the `$0` snippet parts. The example I linked was for how to check whether we want to emit `$0` or not.

**Comment by Arthamys** [comment]
I've go the changes in my fork, would that be sufficient to close the issue ?

https://github.com/Arthamys/rust-analyzer/commit/c118f885dd8f7d13f4b6db12bd622c04a3897f6c

**Comment by Veykril** [comment]
Ye I think that also has to be changed

**Comment by bjorn3** [comment]
Fixed by #7880.
